### ISSUE TITLE
false positive in `reportUnboundVariable`

### ISSUE THREAD

**Issue by tamird** [microsoft/pyright#8218]
```python
@dataclass
class Foo:
    type: str


MAPPING2: Mapping[str, str] = {
    "hello": "world",
}


def func() -> Any:
    return [
        Foo(
            type=(type := "hello"),
            **MAPPING2[type], # error: "type" is unbound (reportUnboundVariable)
        ),
    ]
```

Oddly this is pretty specific to the situation - it doesn't happen if I remove the outer list, or if I replace the construction of `Foo` with a dictionary.

EDIT: This appears to have regressed in `1.1.366`.

**Comment by erictraut** [comment]
This is included in pyright 1.1.369.

**Comment by Andrej730** [comment]
It is 
![image](https://github.com/microsoft/pyright/assets/9417531/44a16aec-67a3-4cc1-86ad-6f04a6390734)
